File: problems/arrays/four_number_sum/js/four_number_sum.js
export const fourNumberSum = (array, targetSum) => {
  array.sort((a, b) => a - b);
  const result = [];

  const skipDuplicates = (arr, index, bound, step) => {
    while (index !== bound && arr[index] === arr[index + step]) {
      index += step;
    }
    if (index !== bound) index += step;
    return index;
  };

  const isDuplicate = (arr, index, bound, prevIndex) => {
    return index > bound && arr[index] === arr[prevIndex];
  };

  for (let i = 0; i < array.length; i++) {
    if (isDuplicate(array, i, 0, i - 1)) continue;

    for (let j = i + 1; j < array.length; j++) {
      if (isDuplicate(array, j, i + 1, j - 1)) continue;

      let left = j + 1;
      let right = array.length - 1;

      while (left < right) {
        const sum = array[i] + array[j] + array[left] + array[right];

        if (sum < targetSum) left++;
        else if (sum > targetSum) right--;
        else {
          result.push([array[i], array[j], array[left], array[right]]);
          left = skipDuplicates(array, left, right, 1);
          right = skipDuplicates(array, right, left, -1);
        }
      }
    }
  }
  return result;
};

File: problems/arrays/four_number_sum/cpp/four_number_sum.cpp
#include "four_number_sum.h"
#include <algorithm>

std::vector<std::vector<int>>
FourNumberSumSolver::fourNumberSum(std::vector<int> array, int targetSum) {
  std::sort(array.begin(), array.end());
  std::vector<std::vector<int>> result;

  auto skipDuplicates = [](const std::vector<int> &array, size_t &index,
                           size_t bound, int step) {
    while (index != bound && array[index] == array[index + step]) {
      index += step;
    }
    if (index != bound)
      index += step;
  };

  auto isDuplicate = [](const std::vector<int> &array, size_t index,
                        size_t bound, size_t prevIndex) {
    return index > bound && array[index] == array[prevIndex];
  };

  for (size_t i = 0; i < array.size(); i++) {
    if (isDuplicate(array, i, 0, i - 1))
      continue;

    for (size_t j = i + 1; j < array.size(); j++) {
      if (isDuplicate(array, j, i + 1, j - 1))
        continue;

      auto left = j + 1;
      auto right = array.size() - 1;

      while (left < right) {
        auto sum = array[i] + array[j] + array[left] + array[right];

        if (sum < targetSum) {
          left++;
        } else if (sum > targetSum) {
          right--;
        } else {
          result.push_back({array[i], array[j], array[left], array[right]});
          skipDuplicates(array, left, right, 1);
          skipDuplicates(array, right, left, -1);
        }
      }
    }
  }
  return result;
}
